version: '3.9'

x-app-args: &app-args
  - USER=${USER}
  - UID=${UID}


#x-app-logging: &app-logging
#  driver: "json-file"
#  options:
#    max-size: "12Mb"
#    max-file: "1"
#
volumes:
  mysql-data:
  elastic-data:
  # mongo-data:
  # kibana-data:
  # rabbitmq-data:
#
networks:
 default:
   driver: bridge
   ipam:
     config:
       - subnet: 11.14.0.0/16
 traefik-net:
   name: traefik_net
   external: true
#
services:
  # Главный сервис бэка
  app-back:
    build:
      context: .
      dockerfile: ./images/php/Dockerfile
      args: *app-args
    depends_on:
      - mysql
      - elasticsearch
    volumes:
      - ${LOCAL_CODE_PATH_BACK}:/var/www
    networks:
      - default
      - traefik-net
  #
  nginx:
    image: nginx:latest
    # ports:
    #   - "80:80"
    #   - "443:443"
    volumes:
      - ./etc/nginx/:/etc/nginx/conf.d/
      - ${LOCAL_CODE_PATH_BACK}:/var/www
    depends_on:
      - app-back
    networks:
      - default
      - traefik-net
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.nginx.entrypoints=web"
      - "traefik.http.routers.nginx.rule=Host(`${LOCAL_HOSTNAME_BACK}`)"
      - "traefik.docker.network=traefik_net"
      # - "traefik.http.services.nginx.loadbalancer.server.port=80"
  #
  mysql:
    image: mysql:8.0
    ports:
      - ${MYSQL_FORWARD_DB_PORT}:3306
    environment:
      MYSQL_ROOT_PASSWORD: ${MYSQL_ROOT_PASSWORD}
    healthcheck:
      test: ["CMD", "mysqladmin" ,"ping", "-h", "localhost", "-uroot", "-p${MYSQL_ROOT_PASSWORD}"]
      interval: 5s
      timeout: 5s
      retries: 20
    ulimits:
      nproc: 65535
      nofile:
        soft: 65535
        hard: 65535
    volumes:
      - mysql-data:/var/lib/mysql
      - ./init/mysql/:/docker-entrypoint-initdb.d/
      - ./etc/mysql/:/etc/mysql/conf.d/
    command: mysqld --sql_mode="ONLY_FULL_GROUP_BY,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION"
    networks:
      - default
    labels:
      traefik.enable: false
  #
  phpmyadmin:
    image: phpmyadmin/phpmyadmin:5.1.3
    # logging: *app-logging
    environment:
      - TZ=Europe/Moscow
      - PMA_HOSTS=mysql
      - PMA_PORT=3306
      - PMA_ABSOLUTE_URI=http://${LOCAL_HOSTNAME_PMA}
    networks:
      - default
      - traefik-net
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=traefik_net"
      # HTTP
      - "traefik.http.routers.phpmyadmin.entrypoints=web"
      - "traefik.http.routers.phpmyadmin.rule=Host(`${LOCAL_HOSTNAME_PMA}`)"
      - "traefik.http.routers.phpmyadmin.middlewares.compress=true"
      # - "traefik.http.services.phpmyadmin.loadbalancer.server.port=80"
  #
  elasticsearch:
    image: elastic/elasticsearch:7.17.12
    environment:
      ES_JAVA_OPTS: "-Xms1024m -Xmx1024m"
      ELASTICSEARCH_USERNAME: "${ELASTIC_USERNAME}"
      ELASTICSEARCH_PASSWORD: "${ELASTIC_PASSWORD}"
      discovery.type: "single-node"
      bootstrap.memory_lock: "true"
      xpack.security.enabled: "false"
      xpack.security.http.ssl.enabled: "false"
      xpack.monitoring.collection.enabled: "true"
    volumes:
      - elastic-data:/usr/share/elasticsearch/data
    ulimits:
      nproc: 65535
      nofile:
        soft: 65535
        hard: 65535
  #
  # mongo-db:
  #   image: mongo:6.0
  #   ports:
  #     - ${LOCAL_FORWARD_MONGO_DB_PORT}:27017
  #   volumes:
  #     - mongo-db-data:/data/db
  #   ulimits:
  #     nproc: 65535
  #     nofile:
  #       soft: 65535
  #       hard: 65535
  #
  # rabbitmq:
  #   image: rabbitmq:3.10-management
  #   healthcheck:
  #     test: rabbitmq-diagnostics -q ping
  #     interval: 30s
  #     timeout: 30s
  #     retries: 3
  #   volumes:
  #     - rabbitmq-data:/var/lib/rabbitmq/mnesia/rabbit@learninglocker-rabbitmq:cached
  #   environment:
  #     - TZ=Europe/Moscow
  #     # - RABBITMQ_DEFAULT_USER=local
  #     # - RABBITMQ_DEFAULT_PASS=local
  #     # - RABBITMQ_DEFAULT_VHOST=/local
  # #
  # redis:
  #   image: redis:7-alpine
  traefik:
    image: traefik:2.10.4
    # logging: *app-logging
    container_name: traefik
    # restart: always
    command:
      - --log.level=ERROR
      - --api.insecure=true
      - --api.dashboard=true
      - --providers.docker=true
      - --entrypoints.web.address=:80
    ports:
      - "80:80"
      # - "8080:8080"
    networks:
      traefik-net:
        aliases:
          - ${LOCAL_HOSTNAME_BACK}
          # - ${LOCAL_HOSTNAME_FRONT}
      default:
        aliases:
          - ${LOCAL_HOSTNAME_BACK}
          # - ${LOCAL_HOSTNAME_FRONT}
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=traefik_net"
      # HTTP
      - "traefik.http.routers.traefik.entrypoints=web"
      - "traefik.http.routers.traefik.rule=Host(`${LOCAL_HOSTNAME_TRAEFIK}`)"
      - "traefik.http.services.traefik.loadbalancer.server.port=8080"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
